provider/azureProvider.js

```diff
@@ -0,0 +1,577 @@
+'use strict';
+
+const BbPromise = require('bluebird');
+const _ = require('lodash');
+const msRestAzure = require('ms-rest-azure');
+const resourceManagement = require('azure-arm-resource');
+const path = require('path');
+const fs = require('fs');
+const fse = require('fs-extra');
+const https = require('https');
+const JSZip = require('jszip');
+const request = require('request');
+const dns = require('dns');
+const parseBindings = require('../shared/parseBindings');
+
+let resourceGroupName;
+let deploymentName;
+let functionAppName;
+let subscriptionId;
+let servicePrincipalTenantId;
+let servicePrincipalClientId;
+let servicePrincipalPassword;
+let functionsAdminKey;
+let invocationId;
+let oldLogs = '';
+let principalCredentials;
+let functionsFolder;
+let existingFunctionApp = false;
+let parsedBindings;
+const zipArray = [];
+const deployedFunctionNames = [];
+
+const constants = {
+  'authorizationHeader': 'Authorization',
+  'bearer': 'Bearer ',
+  'contentTypeHeader': 'Content-Type',
+  'functionAppApiPath': '/api/',
+  'functionAppDomain': '.azurewebsites.net',
+  'functionsAdminApiPath': '.azurewebsites.net/admin/functions/',
+  'functionsApiPath': '/api/functions',
+  'jsonContentType': 'application/json',
+  'logInvocationsApiPath': '/azurejobs/api/functions/definitions/',
+  'logOutputApiPath': '/azurejobs/api/log/output/',
+  'logStreamApiPath': '/api/logstream/application/functions/function/',
+  'masterKeyApiApth': '/api/functions/admin/masterkey',
+  'providerName': 'azure',
+  'scmDomain': '.scm.azurewebsites.net',
+  'scmVfsPath': '.scm.azurewebsites.net/api/vfs/site/wwwroot/',
+  'scmZipApiPath': '.scm.azurewebsites.net/api/zip/site/wwwroot/'
+};
+
+class AzureProvider {
+  static getProviderName () {
+    return constants.providerName;
+  }
+
+  constructor (serverless) {
+    this.serverless = serverless;
+    this.provider = this;
+    this.serverless.setProvider(constants.providerName, this);
+    subscriptionId = process.env[this.serverless.service.provider.subscriptionId];
```

Comment:
This doesn't seem correct. It seems like this used to pull credentials from serverless.yml and then switched to taking them from environment variables, but now instead it kind of does half of both. Environment variables seem better than serverless.yml, because at least it's out of the source, but it's still a worse pattern than the other plugins have. I'm not sure what options there with Azure credentials, though. Anyways, the way the README currently reads, it seems like

```javascript
subscriptionId = process.env['azureSubId'];
servicePrincipalTenantId = process.env['azureServicePrincipalTenantId'];
servicePrincipalClientId = process.env['azureservicePrincipalClientId']; // The 's' should probably get capitalized in the README
servicePrincipalPassword = process.env['azureServicePrincipalPassword'];
```

would be more correct.